src/repo/mod.rs

```diff
@@ -133,35 +133,15 @@ impl<TRepoTypes: RepoTypes> Repo<TRepoTypes> {
     /// Retrives a block from the block store.
     pub async fn get_block(&mut self, cid: &Cid) -> Result<Block, Error>
     {
-        // FIXME: this should probably be a value in the right hand side of some
-        // Map<Cid, Vec<_>> which would always be checked during put_block? Not sure what would
-        // work here, maybe mpsc::oneshot? While unlikely that there is ever many at once.
-        // futures_intrusive would have ManualResetEvent?
-        use futures::compat::Future01CompatExt;
-        use std::time::{Instant, Duration};
-        use tokio::timer::Delay;
-        let mut once = true;
         loop {
             if !self.block_store.contains(&cid).await? {
                 // sending only fails if no one is listening anymore
                 // and that is okay with us.
                 let _ = self.events.send(RepoEvent::WantBlock(cid.clone())).await;
             }
             match self.block_store.get(&cid.clone()).await? {
-                Some(block) => {
-                    if !once {
-                        info!("got block later: {}", cid);
-                    }
-                    return Ok(block);
-                },
-                None => {
-                    if once {
-                        once = false;
-                        info!("getting block later: {}", cid);
-                    }
-                    Delay::new(Instant::now() + Duration::from_millis(5000)).compat().await.unwrap();
-                    continue
-                },
+                Some(block) => return Ok(block),
+                None => continue,
```

Comment:
I think this was broken in the async-std transition? I removed it as it's a hack and won't work without the tokio executor anymore

Reply:
I think this will spin busy without the hack, but yeah it'll need some real solution.

Reply:
So the repo should have a `HashMap<Cid, Vec<Context>>` or something like that.

Reply:
If we migrate to rust-ipld this stuff will change anyway. So the store trait should have a `poll_get` instead of an `async get` and the store should register the future context if the block is not in the store and block. The deamon needs to then send a notify event and the client needs to wake up all the registered contexts for the cid.

Reply:
> `HashMap<Cid, Vec<Context>>`

I think something like this, with something representing the multihash to a list of Wakers could work. Need to check if the latest developments have brought some tools for notifying N tasks at that point.. Can't find from the new primitives in tokio helping here nor immediatedly see how this could be done. Using the receiver side of an one-shot channel wrapped inside a `Arc<Mutex<_>>` to get every task waiting first waiting for mutex, then waiting for a message in turns. Upon completion a sender could be dropped, which would release everyone waiting for messages. That sounds also like a hack :)

Reply:
I was thinking something like this?

```rust
for ctx in contexts {
    ctx.wake()
}
```